src/zenml/stack/stack_component.py

```diff
@@ -63,6 +67,157 @@ class StackComponent(BaseModel, ABC):
+                        "stack component. This is currently only a warning, "
+                        "but future versions of ZenML will require you to pass "
+                        "in senstive information as secrets. Check out the "
+                        "documentation on how to configure your stack "
+                        "components with secrets here: "
+                        "https://docs.zenml.io/developer-guide/advanced-usage/secret-references"
+                    )
+                continue
+
+            if key == "name":
+                raise ValueError(
+                    "Passing the `name` attribute of a stack component as a "
+                    "secret reference is not allowed."
+                )
+
+            requires_validation = field.pre_validators or field.post_validators
+            if requires_validation:
+                raise ValueError(
+                    f"Passing the stack component attribute `{key}` as a "
+                    "secret reference is not allowed as additional validation "
+                    "is required for this attribute."
+                )
+
+        super().__init__(**kwargs)
+
+    def __custom_getattribute__(self, key: str) -> Any:
+        """Returns the (potentially resolved) attribute value for the given key.
+
+        An attribute value may be either specified directly, or as a secret
+        reference. In case of a secret reference, this method resolves the
+        reference and returns the secret value instead.
+
+        Args:
+            key: The key for which to get the attribute value.
+
+        Raises:
+            RuntimeError: If the stack component is not part of the active
+                stack, or the active stack is missing a secrets manager.
+            KeyError: If the secret or secret key don't exist.
+
+        Returns:
+            The (potentially resolved) attribute value.
+        """
+        value = super().__getattribute__(key)
+
+        if not secret_utils.is_secret_reference(value):
+            return value
+
+        from zenml.repository import Repository
+
+        stack = Repository().active_stack
+
+        # A stack component can be part of many stacks, and currently a
+        # secrets manager is associated with a stack. This means we're
+        # not able to identify the 'correct' secrets manager that the user
+        # wanted to resolve the secrets in a general way. We therefore
+        # limit secret resolving to components of the active stack.
+        component = stack.components.get(self.TYPE, None)
+        if not component or component.uuid != self.uuid:
+            raise RuntimeError(
+                f"Failed to resolve secret reference for attribute {key} "
+                f"of stack component `{self}`: The stack component is not "
+                "part of the active stack and therefore can't have it's "
+                "secret references resolved."
```

Comment:
```suggestion
                f"Failed to resolve secret reference for attribute {key} "
                f"of stack component `{self}`: The stack component needs to be "
                "a part of the active stack to have its'  "
                "secret references resolved as the configuration depends
                on the secret manager that is scoped to the respective stack.
                In order to configure this component for a specific stack, set 
                that stack as your active stack and rerun. "
```

I like when Error messages tell me exactly what I can do to achieve what I wanted to do